File: trunk/Geometry/volume.cpp
#include "volume.h"
#include "miscmath.h"

#include <stdio.h>

typedef map<TriPosition, Volume*, TriPositionCompare> VolumeIndex;
static VolumeIndex* Index = NULL;

Volume::Volume( Tetra& t ){
  StdTetra st = labelTetra( t );
    
  len[0] = Length::At( Triangulation::edgeTable[ st.e12 ] );
  len[1] = Length::At( Triangulation::edgeTable[ st.e13 ] );
  len[2] = Length::At( Triangulation::edgeTable[ st.e14 ] );
  len[3] = Length::At( Triangulation::edgeTable[ st.e23 ] );
  len[4] = Length::At( Triangulation::edgeTable[ st.e24 ] );
  len[5] = Length::At( Triangulation::edgeTable[ st.e34 ] );
   
  for(int i = 0; i < 6; i++) len[i]->addDependent( this );        
}

void Volume::recalculate(){
  double CayleyMenger;
  double L12 = len[0]->getValue();
  double L13 = len[1]->getValue();
  double L14 = len[2]->getValue();
  double L23 = len[3]->getValue();
  double L24 = len[4]->getValue();
  double L34 = len[5]->getValue();
   
  CayleyMenger = (-1)*( pow(L12, 4.0)*pow(L34,2.0) + pow(L13, 4.0)*pow(L24,2.0) 
		 + pow(L14, 4.0)*pow(L23,2.0) + pow(L23, 4.0)*pow(L14,2.0)
		 + pow(L24, 4.0)*pow(L13,2.0) + pow(L34, 4.0)*pow(L12,2.0) );

  CayleyMenger += (-1)*( pow(L12, 2.0)*pow(L13,2.0)*pow(L23,2.0) 
			 + pow(L12, 2.0)*pow(L14,2.0)*pow(L24,2.0)
			 + pow(L13, 2.0)*pow(L14,2.0)*pow(L34,2.0)
			 + pow(L23, 2.0)*pow(L24,2.0)*pow(L34,2.0) );

  CayleyMenger += pow(L12, 2.0)*pow(L13,2.0)*pow(L24,2.0) 
                         + pow(L12, 2.0)*pow(L13,2.0)*pow(L34,2.0)
                         + pow(L12, 2.0)*pow(L14,2.0)*pow(L23,2.0) 
                         + pow(L12, 2.0)*pow(L14,2.0)*pow(L34,2.0);

  CayleyMenger += pow(L13, 2.0)*pow(L14,2.0)*pow(L23,2.0) 
                         + pow(L13, 2.0)*pow(L14,2.0)*pow(L24,2.0)
                         + pow(L13, 2.0)*pow(L23,2.0)*pow(L24,2.0)
                         + pow(L14, 2.0)*pow(L23,2.0)*pow(L24,2.0);

  CayleyMenger += pow(L12, 2.0)*pow(L23,2.0)*pow(L34,2.0)
                         + pow(L14, 2.0)*pow(L23,2.0)*pow(L34,2.0) 
                         + pow(L12, 2.0)*pow(L24,2.0)*pow(L34,2.0) 
                         + pow(L13, 2.0)*pow(L24,2.0)*pow(L34,2.0);
   
  value = sqrt(CayleyMenger / 144.0);
}

void Volume::remove() {
   deleteDependents();
   len[0]->removeDependent(this);
   len[1]->removeDependent(this);
   len[2]->removeDependent(this);
   len[3]->removeDependent(this);
   len[4]->removeDependent(this);
   len[5]->removeDependent(this);
   Index->erase(pos);
   delete this;
}

Volume::~Volume(){}

Volume* Volume::At( Tetra& t ){
  TriPosition T( 1, t.getSerialNumber() );
  if( Index == NULL ) Index = new VolumeIndex();
  VolumeIndex::iterator iter = Index->find( T );

  if( iter == Index->end() ){
    Volume* val = new Volume( t );
    val->pos = T;
    Index->insert( make_pair( T, val ) );
    return val;
  } else {
    return iter->second;
  }
}

void Volume::CleanUp(){
  if( Index == NULL ) return;
  VolumeIndex::iterator iter;
  VolumeIndex copy = *Index;
  for(iter = copy.begin(); iter != copy.end(); iter++) {
    iter->second->remove();
  }
    
  delete Index;
  Index = NULL;
}

void Volume::Record( char* filename ){
  FILE* output = fopen( filename, "a+" );

  VolumeIndex::iterator iter;
  for(iter = Index->begin(); iter != Index->end(); iter++)
    fprintf( output, "%lf ", iter->second->getValue() );
  fprintf( output, "\n");

  fclose( output );
}

File: trunk/Java/src/Geoquant/Volume.java
package Geoquant;

import java.util.HashMap;
import java.util.LinkedList;

import Triangulation.StdTetra;
import Triangulation.Tetra;
import Triangulation.Triangulation;

public class Volume extends Geoquant {
  // Index map
  private static HashMap<TriPosition, Volume> Index = new HashMap<TriPosition, Volume>();

  private Length L_ij;
  private Length L_ik;
  private Length L_il;
  private Length L_jk;
  private Length L_jl;
  private Length L_kl;
  
  public Volume(Tetra t) {
    super();
    StdTetra st = new StdTetra(t);
    L_ij = Length.At(st.e12);
    L_ik = Length.At(st.e13);
    L_il = Length.At(st.e14);
    L_jk = Length.At(st.e23);
    L_jl = Length.At(st.e24);
    L_kl = Length.At(st.e34);
    
    L_ij.addDependent(this);
    L_ik.addDependent(this);
    L_il.addDependent(this);
    L_jk.addDependent(this);
    L_jl.addDependent(this);
    L_kl.addDependent(this);
  }
  
  protected void recalculate() {
    double CayleyMenger;
    double L12 = L_ij.getValue();
    double L13 = L_ik.getValue();
    double L14 = L_il.getValue();
    double L23 = L_jk.getValue();
    double L24 = L_jl.getValue();
    double L34 = L_jk.getValue();
     
    CayleyMenger = (-1)*( Math.pow(L12, 4.0)*Math.pow(L34,2.0) + Math.pow(L13, 4.0)*Math.pow(L24,2.0) 
       + Math.pow(L14, 4.0)*Math.pow(L23,2.0) + Math.pow(L23, 4.0)*Math.pow(L14,2.0)
       + Math.pow(L24, 4.0)*Math.pow(L13,2.0) + Math.pow(L34, 4.0)*Math.pow(L12,2.0) );

    CayleyMenger += (-1)*( Math.pow(L12, 2.0)*Math.pow(L13,2.0)*Math.pow(L23,2.0) 
         + Math.pow(L12, 2.0)*Math.pow(L14,2.0)*Math.pow(L24,2.0)
         + Math.pow(L13, 2.0)*Math.pow(L14,2.0)*Math.pow(L34,2.0)
         + Math.pow(L23, 2.0)*Math.pow(L24,2.0)*Math.pow(L34,2.0) );

    CayleyMenger += Math.pow(L12, 2.0)*Math.pow(L13,2.0)*Math.pow(L24,2.0) 
                           + Math.pow(L12, 2.0)*Math.pow(L13,2.0)*Math.pow(L34,2.0)
                           + Math.pow(L12, 2.0)*Math.pow(L14,2.0)*Math.pow(L23,2.0) 
                           + Math.pow(L12, 2.0)*Math.pow(L14,2.0)*Math.pow(L34,2.0);

    CayleyMenger += Math.pow(L13, 2.0)*Math.pow(L14,2.0)*Math.pow(L23,2.0) 
                           + Math.pow(L13, 2.0)*Math.pow(L14,2.0)*Math.pow(L24,2.0)
                           + Math.pow(L13, 2.0)*Math.pow(L23,2.0)*Math.pow(L24,2.0)
                           + Math.pow(L14, 2.0)*Math.pow(L23,2.0)*Math.pow(L24,2.0);

    CayleyMenger += Math.pow(L12, 2.0)*Math.pow(L23,2.0)*Math.pow(L34,2.0)
                           + Math.pow(L14, 2.0)*Math.pow(L23,2.0)*Math.pow(L34,2.0) 
                           + Math.pow(L12, 2.0)*Math.pow(L24,2.0)*Math.pow(L34,2.0) 
                           + Math.pow(L13, 2.0)*Math.pow(L24,2.0)*Math.pow(L34,2.0);
     
    value = Math.sqrt(CayleyMenger / 144.0);

  }

  protected void remove() {
    deleteDependents();
    L_ij.removeDependent(this);
    L_ik.removeDependent(this);
    L_il.removeDependent(this);
    L_jk.removeDependent(this);
    L_jl.removeDependent(this);
    L_kl.removeDependent(this);
    Index.remove(pos);
  }

  public static Volume At(Tetra t) {
    TriPosition T = new TriPosition(t.getSerialNumber());
    Volume q = Index.get(T);
    if(q == null) {
      q = new Volume(t);
      q.pos = T;
      Index.put(T, q);
    }
    return q;
  }
  
  public static double valueAt(Tetra t) {
    return At(t).getValue();
  }
  
  public static class Sum extends Geoquant {
    LinkedList<Volume> volumes = new LinkedList<Volume>();
    private static Sum total = null;
    
    private Sum() {
      super();
      Volume v;
      for(Tetra t : Triangulation.tetraTable.values()) {
        v = Volume.At(t);
        v.addDependent(this);
        volumes.add(v);
      }
    }
    protected void recalculate() {
      value = 0;
      for(Volume v : volumes) {
        value += v.getValue();
      }
    }

    protected void remove() {
      deleteDependents();
      for(Volume v : volumes) {
        v.removeDependent(this);
      }
      volumes.clear();
    }
    
    public static Sum get() {
      if(total == null) {
        total = new Sum();
      }
      return total;
    }
    
    public static double getSum() {
      return total.getValue();
    }
  }

}
